cpp/src/io/json/data_casting.cuh

```diff
@@ -0,0 +1,389 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <io/utilities/parsing_utils.cuh>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/strings/detail/utf8.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cub/cub.cuh>
+
+#include <memory>
+
+namespace cudf::io::json::experimental {
+
+// Unicode code point escape sequence
+static constexpr char UNICODE_SEQ = 0x7F;
+
+// Invalid escape sequence
+static constexpr char NON_ESCAPE_CHAR = 0x7E;
+
+// Unicode code point escape sequence prefix comprises '\' and 'u' cahrs
+static constexpr size_type UNICODE_ESC_PREFIX = 2;
+
+// Unicode code point escape sequence comprises four hex characters
+static constexpr size_type UNICODE_HEX_DIGIT_COUNT = 4;
+
+// A unicode code point escape sequence is \uXXXX
+static auto constexpr NUM_UNICODE_ESC_SEQ_CHARS = UNICODE_ESC_PREFIX + UNICODE_HEX_DIGIT_COUNT;
+
+static constexpr auto UTF16_HIGH_SURROGATE_BEGIN = 0xD800;
+static constexpr auto UTF16_HIGH_SURROGATE_END   = 0xDC00;
+static constexpr auto UTF16_LOW_SURROGATE_BEGIN  = 0xDC00;
+static constexpr auto UTF16_LOW_SURROGATE_END    = 0xE000;
+
+/**
+ * @brief Returns the character to output for a given escaped character that's following a
+ * backslash.
+ *
+ * @param escaped_char The character following the backslash.
+ * @return The character to output for a given character that's following a backslash
+ */
+__device__ __forceinline__ char get_escape_char(char escaped_char)
+{
+  switch (escaped_char) {
+    case '"': return 0x22;
+    case '\\': return 0x5C;
+    case '/': return 0x2F;
+    case 'b': return 0x08;
+    case 'f': return 0x0C;
+    case 'n': return 0x0A;
+    case 'r': return 0x0D;
+    case 't': return 0x09;
+    case 'u': return UNICODE_SEQ;
+    default: return NON_ESCAPE_CHAR;
+  }
+}
```

Comment:
1. This seems to be an incomplete list of escape characters: `\'` is missing. Also, we are missing forms like `\xFF` (for one-byte hexadecimal escapes), octals `\XXX`, and Unicode code point escapes `\u{XXXXX}`. I'm not sure which of these we care to support.

https://en.wikipedia.org/wiki/Escape_character#JavaScript
https://mathiasbynens.be/notes/javascript-escapes
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Grammar_and_types#using_special_characters_in_strings

2. We don't need to use hexadecimal identifiers for the characters here. We can just use the raw character or the C++ escape sequence.
```suggestion
__device__ __forceinline__ char get_escape_char(char escaped_char)
{
  switch (escaped_char) {
    case '\'': return '\'';
    case '"': return '"';
    case '\\': return '\\';
    case '/': return '/';
    case 'b': return '\b';
    case 'f': return '\f';
    case 'n': return '\n';
    case 'r': return '\r';
    case 't': return '\t';
    case 'u': return UNICODE_SEQ;
    default: return NON_ESCAPE_CHAR;
  }
}
```

https://godbolt.org/z/vMrn5ME1r

Reply:
Thanks, @bdice. With regard to the set of escape sequences, we had decided to follow the [ECMA-404](https://www.ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf#page=12) and [RFC 7159](https://www.ietf.org/rfc/rfc7159.txt).

Reply:
Thanks, I've adopted the suggestion of representing the return values as char literals. 👍 

I kept the set of supported escape sequences as mentioned above to [ECMA-404](https://www.ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf#page=12) and [RFC 7159](https://www.ietf.org/rfc/rfc7159.txt). Let me know if you think we should reconsider and include the ones mentioned for JavaScript, otherwise feel free to resolve.

Reply:
That's fine with me! If we have some standard that we like, I'm happy.